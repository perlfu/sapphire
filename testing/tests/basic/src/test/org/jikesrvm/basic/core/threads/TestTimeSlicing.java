/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

class TestTimeSlicing {

  private static final int LAST_COUNT = 10; // threads count to ten in step

  public static void main(String[] args) {

    final Task[] tasks = new Task[3];
    final int[] countArray = new int[tasks.length];

    for (int i = 0; i < tasks.length; ++i) {
      tasks[i] = new Task(i, countArray);
    }

    //start up tasks
    for (Task task : tasks) { task.start(); }

    // spin and wait for values to fill
    while (countArray[tasks.length - 1] != LAST_COUNT) {
      spinner();
    }

    XThread.say("bye");
    XThread.outputMessages();
  }

  // static routine to spin a bit
  //
  static int spinner() {
    // dummy calculation
    int sum = 0;
    for (int i = 1; i < 100; i++) {
      sum += i;
    }
    Thread.yield();
    return sum;
  }

  static class Task extends XThread {
    private final int id;
    private final int[] values;
    private final int previous;


    //---------------------------------

    Task(int id, int[] values) {
      super("Task " + id);
      this.id = id;
      this.values = values;
      this.previous = (id + values.length - 1) % values.length;
    }

    void performTask() {
      while (values[id] < LAST_COUNT) {
        boolean increment = false;
        if( 0 == id && values[previous] == values[id] ) {
          increment = true;
        } else if( 0 != id && values[previous] - 1 == values[id] ) {
          increment = true;
        }

        if( increment ) {
          values[id]++;
          tsay("incremented to count: " + values[id]);
        }
        //  spin for a while
        spinner();  // virtual spin code
      }
    }
  }
}
