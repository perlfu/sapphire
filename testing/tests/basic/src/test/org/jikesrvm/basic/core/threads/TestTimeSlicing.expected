Thread 0 Main: bye
Thread 1 Task 0: creating
Thread 1 Task 0: starting
Thread 1 Task 0: run starting
Thread 1 Task 0: incremented to count: 1
Thread 1 Task 0: incremented to count: 2
Thread 1 Task 0: incremented to count: 3
Thread 1 Task 0: incremented to count: 4
Thread 1 Task 0: incremented to count: 5
Thread 1 Task 0: incremented to count: 6
Thread 1 Task 0: incremented to count: 7
Thread 1 Task 0: incremented to count: 8
Thread 1 Task 0: incremented to count: 9
Thread 1 Task 0: incremented to count: 10
Thread 1 Task 0: run finishing
Thread 2 Task 1: creating
Thread 2 Task 1: starting
Thread 2 Task 1: run starting
Thread 2 Task 1: incremented to count: 1
Thread 2 Task 1: incremented to count: 2
Thread 2 Task 1: incremented to count: 3
Thread 2 Task 1: incremented to count: 4
Thread 2 Task 1: incremented to count: 5
Thread 2 Task 1: incremented to count: 6
Thread 2 Task 1: incremented to count: 7
Thread 2 Task 1: incremented to count: 8
Thread 2 Task 1: incremented to count: 9
Thread 2 Task 1: incremented to count: 10
Thread 2 Task 1: run finishing
Thread 3 Task 2: creating
Thread 3 Task 2: starting
Thread 3 Task 2: run starting
Thread 3 Task 2: incremented to count: 1
Thread 3 Task 2: incremented to count: 2
Thread 3 Task 2: incremented to count: 3
Thread 3 Task 2: incremented to count: 4
Thread 3 Task 2: incremented to count: 5
Thread 3 Task 2: incremented to count: 6
Thread 3 Task 2: incremented to count: 7
Thread 3 Task 2: incremented to count: 8
Thread 3 Task 2: incremented to count: 9
Thread 3 Task 2: incremented to count: 10
Thread 3 Task 2: run finishing
